src/Text.Analyzers/Core (Net4)/IdentifiersShouldBeSpelledCorrectly.cs

```diff
@@ -0,0 +1,416 @@
+            MethodTypeParameterRule,
+            AssemblyMoreMeaningfulNameRule,
+            NamespaceMoreMeaningfulNameRule,
+            TypeMoreMeaningfulNameRule,
+            MemberMoreMeaningfulNameRule,
+            MemberParameterMoreMeaningfulNameRule,
+            DelegateParameterMoreMeaningfulNameRule,
+            TypeTypeParameterMoreMeaningfulNameRule,
+            MethodTypeParameterMoreMeaningfulNameRule
+        );
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            // Do everything within a CompilationStartAction so we have access to the assembly name before the other actions run
+            analysisContext.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                var assemblyName = compilationStartContext.Compilation.AssemblyName;
+
+                // Store the assembly name each time this analyzer runs so the appropriate custom dictionaries can be retrieved from the cache.
+                // The cache will remain in memory until Visual Studio is closed, so it's possible multiple solutions containing the same
+                // assembly names will be opened while the cache persists. This is not a problem because the cache associating assembly name
+                // to dictionaries is refreshed on the call to LoadDictionaries.
+                _namingService = new NamingService(assemblyName);
+                _namingService.LoadDictionaries(compilationStartContext.Options.AdditionalFiles, compilationStartContext.CancellationToken);
+
+                // CompilationStartAction doesn't support reporting diagnostics, so we have to check the assembly name in CompilationEndAction
+                compilationStartContext.RegisterCompilationEndAction(compilationEndContext =>
+                {
+                    if (assemblyName.Length == 1)
+                    {
+                        var diagnostic = Diagnostic.Create(AssemblyMoreMeaningfulNameRule, Location.None, assemblyName);
+                        compilationEndContext.ReportDiagnostic(diagnostic);
+                    }
+                    else
+                    {
+                        foreach (var misspelledWord in GetMisspelledWords(assemblyName))
+                        {
+                            var diagnostic = Diagnostic.Create(AssemblyRule, Location.None, misspelledWord, assemblyName);
+                            compilationEndContext.ReportDiagnostic(diagnostic);
+                        }
+                    }
+                });
+
+                // Find IdentifierTokens with the syntax tree because SyntaxNodeAction only filters on "methods and above"
+                // Reference: http://stackoverflow.com/a/28004943
+                compilationStartContext.RegisterSyntaxTreeAction(syntaxTreeContext =>
```

Comment:
This is wrong. `RegisterSyntaxTreeAction` is designed only for reporting syntax diagnostics, where semantic model is not required. If you want to use the semantic model, you should register a semantic model action.

Reply:
What's the difference between my approach and `RegisterSemanticModelAction`'s `context.SemanticModel.SyntaxTree.GetRoot().DescendantToken()`? While I want to use the semantic model, I also want to examine syntax tokens.

Reply:
Syntax tree actions are assumed to be lightweight which do not lead to any binding (expensive computation). This is important especially in the IDE live analysis where the IDE analyzer driver front loads all the syntax tree actions (including compiler's syntax/parse diagnostics).

Reply:
Ahh, thanks for the clarification. Now I need to change another internal analyzer  ;)

Reply:
Switched to Symbol/Operation actions